### PR TITLE
Reorganize signup/login modal code

### LINKED ISSUE #4747: [urgent] Login page link to signup form does not work
At https://publiclab.org/login, if you click the `sign up to join` link, nothing happens -- i think it's supposed to work in the modal but is not working on this page.

i'm tracking another login issue, related to OpenID, but will circle back to this.

### PR DESCRIPTION
**Pull request by jywarren**
Tidying up the modal code a bit @SidharthBansal ! 👍 

I'm also interested in understanding it better. I read this: https://github.com/publiclab/plots2/blob/master/doc/LOGIN_SYSTEMS.md#how-to-setup-login-modal-on-various-locations

But when reading this code, I am not sure how `$.getJSON()` works. Why do we fetch the link in JSON format, and actually why do we fetch it at all?

https://github.com/publiclab/plots2/blob/7bb6fac8909690e7a5d7c611824d944285c69a61/app/views/user_sessions/new.html.erb#L44-L55

I would have guessed we'd have a hidden input, like this:

```html
<input type="hidden" name="redirect_to" id="login-modal-redirect" value="<%= request.fullpath %>" />
```

And then be able to assign new values to it, like:

```js
$('#login-modal-redirect').val('/some/new/url'); // rework the form to change the redirect destination
```

Why do we use this other method, just curious? @Dsxv -- i see it looks like you wrote this in #4605, can you walk me through it a bit? 

https://github.com/publiclab/plots2/pull/4605/files#diff-3c3ba86760706389841cd20ee62eef16R47

What I'm wondering is if, for example, @IshaGupta18 is now implementing a method to change the redirect from inside a popover (which is tough because the popover doesn't render its HTML until you activate it, so it's missed by the `requireLogin` jQuery here), it'd be nice to be able to just display the modal simply and then just update the hidden `<input>` as in my code above. Do you think it's worth refactoring for something a bit simpler like that?

But also perhaps I'm just missing how it works and am not understanding! Thanks a lot for helping me walk through this. 

Fixes #4747

### CODE REVIEW

**Comment by jywarren** [comment]
Well at least Dangerbot is working!!! 😂

**Comment by jywarren** [comment]
OK, i made an attempt at implementing it the other way here, using a hidden input: 

https://github.com/publiclab/plots2/pull/4739/commits/10ea71aab0403923d75224c216e70a7dd8f1ecc5

**Comment by jywarren** [comment]
And added a method called `require_login_and_redirect_to(url)`!

**Comment by jywarren** [comment]
I'm also trying to resolve the persistent issue of the Recaptcha covering the button... ugh!

image

**Comment by jywarren** [comment]
So my goals here are:

* [ ] recaptcha fix
* [ ] get "or sign up to join" link to work on /login page
* [ ] refactor to create a `require_login_and_redirect(url)` function (done but needs testing)
* [ ] fix modal from not showing up (recently encountered, not sure why)
* [ ] ensure it's possible to go from signup page to login page by clicking Login

**Comment by jywarren** [comment]
OK, i think the `return_to` should work on signup in that it should show you a link saying "pick up where you left off", here:

https://github.com/publiclab/plots2/blob/a841af031dd54c578b6f1a152fcdb8d16e19beeb/app/controllers/users_controller.rb#L27-L28

However, let's think about how we want them to be signed up. New users shouldn't be dropped off on the subscriptions page. Maybe we should... detect the `return_to=/subscribe/multiple/tag/arduino,games` format and intercept it? Say something else?

Could we match `/subscribe/multiple/tag/arduino,games` and actually parse out the tagnames to subscribe them automatically? What's best here?

Also, i have NO idea about the recaptcha thing. We should try messing with it on https://unstable.publiclab.org using the Chrome dev console to try to get it to not overlap... mess with the CSS. We don't have to solve that here, just not forget to fix it.

**Comment by jywarren** [comment]
`require_login_and_redirect_to('/subscribe/multiple/' + checkedTags);` works pretty well and we can probably open new issues for the Recaptcha and getting it to work on signup as well as login. What do you think, @SidharthBansal and @IshaGupta18 ?

**Comment by jywarren** [comment]
we should push it to unstable again with this latest version and test thoroughly

**Comment by jywarren** [comment]
OK, pushing to unstable -- but also confirming that the `return_to` also worked on signup -- i see:

`<div class="alert alert-success"><button type="button" class="close" data-dismiss="alert">×</button>Registration successful. Welcome to our community!  Click <a href='/subscribe/multiple/tag/arduino,games'>here</a> to continue where you left off.</div>`

**Comment by jywarren** [comment]
OK, pushed. if anyone has a chance to try logging in or signing up from a page like this: http://unstable.publiclab.org/licenses

Then confirming that when you do, you are returned to that page, that would be great!

**Comment by jywarren** [comment]
If you want, in the JavaScript console, you should also be able to run `require_login_and_redirect_to('/licenses')` or any other URL and it should send you there after you log in.

**Comment by IshaGupta18** [comment]
Oh wow, this is just amazing work done here. I think you have solved all the troubles we were facing here in this PR. We could open up a new issue for recaptcha, but since I tried to solve it before, I realized that it's not that easy to correct through dev console, so we could probably test it on unstable and rectify it there. What do you think @jywarren ? Also, please let me know if you need any other help with this issue, you have done it really well!

**Comment by SidharthBansal** [comment]
Jeff you did a lot of stuff in these files. Awesome work.

File: app/views/users/_create_form.html.erb [hunk 1]
```diff
@@ -71,14 +77,16 @@
       %>
     </div>
 
-      <% if Rails.env == "production" && !params[:spamaway] %>
-        <div style="height: 120px">
-          <%= recaptcha_tags %>
-        </div>
+    <% if Rails.env == "production" && !params[:spamaway] %>
+      <div class="col-sm-12 recaptcha-container">
```
Comment (jywarren) [inline app/views/users/_create_form.html.erb]: HAHA fixed the Recaptcha issue -- it needed this class `col-sm-12`:

image

**Comment by jywarren** [comment]
I think this is good to merge. 2 follow-ups will have to include:

* [ ] ability to go directly to `/subscribe/multiple/tag/tag1,tag2` and correctly enter and login (let's open an issue, it should be working due to https://github.com/publiclab/plots2/blob/5cde763bbdc3bfb9b91c26d68389285050b5ba8c/app/controllers/subscription_controller.rb#L178-L179 and https://github.com/publiclab/plots2/blob/5cde763bbdc3bfb9b91c26d68389285050b5ba8c/app/controllers/user_sessions_controller.rb#L142 but it isn't on unstable
* [ ] ability to sign up and get subscribed (see discussion above), opening in https://github.com/publiclab/plots2/issues/4759
* [ ] document `require_login_and_redirect_to(relative_path)` and use it in #4675!

**Comment by jywarren** [comment]
Merging this now. @IshaGupta18 would you want to document the new JavaScript function above and pick back up in #4675? Probably here: https://github.com/publiclab/plots2/blob/b1c57446d016f8cd0ec149a75298711270e1643e/doc/LOGIN_SYSTEMS.md#how-to-setup-login-modal-on-various-locations

**Comment by SidharthBansal** [comment]
This is a lot of work. Thanks both of you.

Jeff I am sorry. I was not able to help you through this. I was really
busy. University has started and it is sometimes very difficult to complete
practicals, vivas and assignment with PL commitment. I can't find time to
be online past week.

Apologies. I know I have been very inactive this one week. Will try to find
time and catch up things. Will try to do both PL and university things
again.

**Comment by IshaGupta18** [comment]
Yes @jywarren I will document this, however, I will need just a bit of time in what exactly you have done and what needs to be done in #4675 . This is actually a bit stressful week for me that's all. Amazing work here and thanks a ton!

**Comment by jywarren** [comment]
no problem - we should just highlight how to use this new JS function, in
the docs, and then we can actually apply it in the subscribe popover!

Good luck with your work, don't overdo it!

**Pull request merged**